Reworked Railgun

BluSunrize opened the pull request:
- Changes the Railgun API around a bunch
- Buffs its damage
- Changes the model
- Allows more variable ammo

WIP, still want to rework the ammo switching

BluSunrize committed:
Change railgun model

BluSunrize committed:
Adjust all railgun shader textures to match the new model

BluSunrize committed:
Allow railgun to cycle through ammo with a keybind, change HUD display for charging

BluSunrize committed:
Allow railgun projectile properties to override the projectile entity

BluSunrize committed:
Allow sawblades to be used as railgun ammo

BluSunrize committed:
Allow blazerods as railgun ammo

BluSunrize committed:
Add a chance for railgun projectiles to break

BluSunrize committed:
Buff damage for railgun

BluSunrize committed:
Fix normals on the railgun model

BluSunrize committed:
Rewrite the Railgun API to be far more variable
We're doing a major version bump soon anyway

BluSunrize committed:
Allow tridents as railgun ammo

BluSunrize committed:
Change railgun to zoom with middle-mouse instead of holding crouch

BluSunrize committed:
Change railgun to switch ammo on sneak + scroll

BluSunrize committed:
Quintuple the energy storage of the railgun

BluSunrize committed:
Reflavor chargetime upgrade for the railgun to be "heatsinks", adjust railgun and upgrade recipes

BluSunrize committed:
Update manual entry for the railgun

BluSunrize committed:
Add an advancement for upgrading the railgun

BluSunrize commented:
@SkySom @malte0811 
This is ready for review at your earliest convenience

BluSunrize committed:
Fix piercing projectiles not keeping track of their pierce counter

malte0811 commented on src/main/java/blusunrize/immersiveengineering/api/tool/RailgunHandler.java:
```diff
@@ -1,70 +1,148 @@
+		/**
+		 * @param shooter
+		 * @param ammo
+		 * @param defaultProjectile
```
I would either remove the `@param` declarations or add documentation for them.

BluSunrize replied:
Will do

malte0811 commented on src/main/java/blusunrize/immersiveengineering/api/tool/RailgunHandler.java:
```diff
@@ -1,70 +1,148 @@
+
+		/**
+		 * @return whether this projectile should be allowed to be placed in, and used from a turret.<br>
+		 * Projectiles that rely on a player using them should return false, since turrets parse null for a player
```
`parse` -> `pass`?

BluSunrize replied:
Yup, typo'd

malte0811 commented on src/main/java/blusunrize/immersiveengineering/api/tool/RailgunHandler.java:
```diff
@@ -1,70 +1,148 @@
-		public int[][] colourMap = {{0x686868, 0xa4a4a4, 0xa4a4a4, 0xa4a4a4, 0x686868}};
+		private double damage;
+		private double gravity;
+		private int[][] colourMap = {{0x686868, 0xa4a4a4, 0xa4a4a4, 0xa4a4a4, 0x686868}};
```
It might be a good idea to use a custom object here, rather than having to guess what which color is used for.

BluSunrize replied:
I'll look into it and see what I come up with

malte0811 commented on src/main/java/blusunrize/immersiveengineering/common/IEContent.java:
```diff
@@ -841,10 +851,57 @@ public static void init()
+				}
+				return defaultProjectile;
+			}
+		});
```
I would suggest moving this to a different class, `IEContent` is too big as it is already.

BluSunrize replied:
I suppose we have initialization of bullets in a separate class too. I'll see to it.

malte0811 commented on src/main/java/blusunrize/immersiveengineering/common/entities/IEProjectileEntity.java:
```diff
@@ -43,6 +44,7 @@
 	public boolean inGround;
 	public int ticksInGround;
 	public int ticksInAir;
+	protected IntOpenHashSet piercedEntities;
```
Why not just `IntSet` for the declaration? And why aren't you just initializing it here or in the constructor?

BluSunrize replied:
This is how vanilla does it. I just didn't bother AT'ing their set, because I'd then also need to AT the list of hit entities (dunno why they have two storages).
But in style with using the most generic interface to refer to fields, I suppose I can switch this to IntSet!

BluSunrize committed:
Fix documentation and field access in RailgunHandler

BluSunrize committed:
Change colouration system for railgun projectiles

BluSunrize committed:
Move railgun projectile registration to separate class

BluSunrize committed:
Change pieced entity set type to be IntSet
We don't initialize it on construction, because its size depends on the piece level of the projectile

BluSunrize commented:
Ready for another review

SkySom commented on src/main/java/blusunrize/immersiveengineering/common/items/RailgunProjectiles.java:
```diff
@@ -0,0 +1,108 @@
+	public static void register()
+	{
+		// Iron
+		RailgunHandler.registerStandardProjectile(Ingredient.fromTag(IETags.ironRod), 16, 1.25).setColourMap(
```
Lack of consistency in spelling of Color/Colour

BluSunrize replied:
>_>
<_<

SkySom commented on src/main/java/blusunrize/immersiveengineering/common/data/Advancements.java:
```diff
@@ -216,6 +216,16 @@ public void accept(Consumer<Advancement> consumer)
 					.withCriterion("railgun", InventoryChangeTrigger.Instance.forItems(Weapons.railgun))
 					.register(consumer, "immersiveengineering:main/craft_railgun");
 
+			ItemStack upgraded_railgun = new ItemStack(Weapons.railgun);
```
_ in a non constant variable?

BluSunrize replied:
Sorry, been writing a lot of python recently. Underscore naming is our variable pattern at work =P

BluSunrize committed:
Fix color naming consistency for the railgun

BluSunrize committed:
Fix underscored variable names to be camelcase

BluSunrize commented:
Awaiting next (hopefully final) review =P
